rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -477,6 +477,62 @@ extern void ppu_register_function_at(u32 addr, u32 size, ppu_function_t ptr)
 	}
 }
 
+extern void ppu_register_function_at(u32 addr, u32 size, u64 ptr)
+{
+	return ppu_register_function_at(addr, size, reinterpret_cast<ppu_function_t>(ptr));
+}
+
+struct ppu_far_jumps_t
+{
+	std::unordered_map<u32, u32> vals;
+
+	shared_mutex mutex;
+};
+
+static bool ppu_far_jump(ppu_thread& ppu)
+{
+	ppu.cia = g_fxo->get<const ppu_far_jumps_t>().vals.at(ppu.cia);
+	return false;
+}
+
+bool ppu_form_branch_to_code(u32 entry, u32 target)
+{
+	entry &= -4;
+	target &= -4;
+
+	if (!vm::check_addr(entry, vm::page_executable) || !vm::check_addr(target, vm::page_executable))
+	{
+		return false;
+	}
+
+	g_fxo->init<ppu_far_jumps_t>();
+
+	auto& inst = *vm::get_super_ptr<u32>(entry);
+
+	if (entry - target <= 0x2000000 || target - entry < 0x2000000)
+	{
+		// Use relative branch
+		inst = ppu_instructions::B(target - entry);
+	}
+	else if (target < 0x2000000)
+	{
+		// Use absolute branch if possible
+		inst = ppu_instructions::B(target, true);
+	}
+	else
+	{
+		// Use special 32-bit target made-up branch
+		// Use illegal opcode for terminating analyzer blocks before it
+		inst = 0;
```

Comment:
It seems there is no need to use illegal opcode at all... PPU Analyser can handle invisible branch using information provided by patch engine, not by treating specific opcode as special. Maybe by checking "ppu far jump" table. Also could possibly use this codepath in all cases instead of B instruction as well.

Reply:
Fixed.